### PR TITLE
REST API: Magic Link for WordPress.com authentication, Part 1

### PR DESCRIPTION
**Pull request by hichamboushaba**
Part 1 of #8427 

⚠️ Please don't merge until the FluxC part is merged: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2667

### Description
This PR adds the first part of the Magic Link flow: the magic link request part.
The second part of handling the magic link will be added in a separate PR.

##### PasswordLess account
1. If you don't have a passwordless account, you can create one either by using the signup flow in the WordPress app, or by creating a new account in the web using social login.
2. Open the app, then enter the address of a non-Jetpack site.
3. Enter site credentials.
4. In the dashboard, click on the Jetpack benefits.
5. Click on "Login to continue"
6. Enter your email address.
7. Confirm that the Magic Link screen is opened, and that a magic link is requested automatically.
8. After the request is finished, confirm that the content is updated to reflect this.
9. Put a breakpoint in this [line](https://github.com/woocommerce/woocommerce-android/blob/26ca3e1ef4855ef383b76ded01bbb48386b6dc50/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/LoginActivity.kt#L257), and start debugging the app.
10. Open the email in the phone, then click on the link in the email you received.
11. Confirm in the debug console that the link has a `flow` and `source` values.

##### Account with password
1. Open the app, then enter the address of a non-Jetpack site.
2. Enter site credentials.
3. In the dashboard, click on the Jetpack benefits.
4. Click on "Login to continue"
5. Enter your email address.
6. In the Password screen, click on "Or continue using Magic Link"
7. Confirm that the Magic Link screen is opened, and that a magic link is requested automatically.
8. Confirm that the "Enter your password instead" button is present.
9. Test the button and confirm it takes you back to the password screen. 
10. The rest is similar to the first TC.

### Images/gif

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by hichamboushaba** [comment]
The `buildkite/woocommerce-android/dependency-tree-diff` is failing because of the merge conflict, I'll fix it after the PR is reviewed to avoid doing the work twice.

**Comment by hafizrahman** [comment]
Both cases work correctly in my test. 

For the first case, I can confirm that the value for `flow` is `sitecredentials-to-wpcom` and for `source` it's `jetpack-connection`.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkProperties.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.woocommerce.android.ui.login
+
+import org.wordpress.android.fluxc.store.AccountStore.AuthEmailFlow
+import org.wordpress.android.fluxc.store.AccountStore.AuthEmailSource
+
+enum class MagicLinkSource(private val value: String) : AuthEmailSource {
+    JetpackInstallation("jetpack-installation"),
+    JetpackConnection("jetpack-connection"),
+    WPComAuthentication("wpcom-authentication\n");
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkProperties.kt]: ❓  What is the purpose for the newline character?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationMagicLinkRequestFragment.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                is Exit -> findNavController().navigateUp()
+            }
+        }
+    }
+
+    private fun openEmailClient() {
+        if (ActivityUtils.isEmailClientAvailable(requireContext())) {
+            ActivityUtils.openEmailClient(requireContext())
+        } else {
+            uiMessageResolver.showSnack(R.string.login_email_client_not_found)
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationMagicLinkRequestFragment.kt:63]: Not related to this PR:

This string currently says "Can't detect your email client app".

I wonder if it's better to make it more actionable, something like "Can't detect email client app. Please install an email app and check your email with it."

**Review by hafizrahman — APPROVED** [review]
Code looks good, great work. Works as expected as well.

Just left a couple of non-blocking comments.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkProperties.kt (hunk 1 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkProperties.kt]: Good catch, it was just a typo.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationMagicLinkRequestFragment.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationMagicLinkRequestFragment.kt:63]: I think a long text like this won't fit in a Snackbar, and also this a rare edge case, all phones ship with an email client already installed.

**Pull request merged**